common/scala/src/main/scala/whisk/core/database/DocumentFactory.scala

```diff
@@ -135,13 +139,13 @@ trait DocumentFactory[W <: DocumentRevisionProvider] extends MultipleReadersSing
       implicit val ec = db.executionContext
 
       val key = CacheKey(doc)
-      val docInfo = doc.docinfo
       val src = StreamConverters.fromInputStream(() => bytes)
 
-      cacheUpdate(doc, key, db.attach(docInfo, attachmentName, contentType, src) map { newDocInfo =>
-        doc.revision[W](newDocInfo.rev)
-        doc.docinfo
-      })
+      db.putAndAttach[W](doc, update, contentType, src, oldAttachment) flatMap {
+        case (newDocInfo, attached) =>
+          val newDoc = update(doc, attached).revision[W](newDocInfo.rev)
+          cacheUpdate(newDoc, key, Future.successful(newDocInfo))
+      }
```

Comment:
Had to change the way cache is updated. So far the cache entry was created in `DocumentFactory` at start of `put` with the document set as value and a future registered to updates its revision post db operation gets completed. This follows the similar patterns as in get and delete. This worked so far as the document which is added just needs the revision to be updated. 

With `putAndAttach` we need to update the `exec` also and thus its not possible to follow current approach as `exec` is immutable (compared to `rev` which is mutable). So now the cache is updated *after* db operation gets completed. 

From what I understand this may ok for put as current approach is more useful for lookups where intention is to ensure that first request that arrives for a certain cache key causes a future to be put into the cache which all later requests then “hook into”